src/weekly/mod.rs

```diff
@@ -0,0 +1,411 @@
+                measured = measured.saturating_add(last_length);
+                last_range = win.range.clone();
+            } else {
+                // Overlapping window, coalesce into the last range.
+                last_range.end = u32::max(last_range.end, win.range.end);
+            };
+        }
+        // Account for the still pending length of the last range.
+        let last_length = last_range.end.saturating_sub(last_range.start + 1);
+        measured = measured.saturating_add(last_length);
+
+        u64::from(measured)
+    }
+
+    /// Return total length of all windows in the calendar.
+    ///
+    /// In case of overlapping windows, total length can be larger than the
+    /// actual amount of weekly minutes in the calendar.
+    #[cfg(test)]
+    pub fn total_length_minutes(&self) -> u64 {
+        self.windows.iter().fold(0u64, |len, win| {
+            len.saturating_add(win.value.length_minutes().into())
+        })
+    }
+
+    /// Return all weekly windows (if any) which contain a given datetime.
+    #[cfg(test)]
+    pub fn containing_windows(&self, datetime: &DateTime<Utc>) -> Vec<&WeeklyWindow> {
+        let timepoint = utils::datetime_as_weekly_minute(datetime);
+        self.windows
+            .query_point(timepoint)
+            .map(|elem| &elem.value)
+            .collect()
+    }
+}
+
+impl Serialize for WeeklyCalendar {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        use serde::ser::SerializeSeq;
+
+        let len = self.windows.iter().count();
+        let mut seq = serializer.serialize_seq(Some(len))?;
+        for interval in self.windows.iter() {
+            seq.serialize_element(&interval.value)?;
+        }
+        seq.end()
+    }
+}
+
+impl Default for WeeklyCalendar {
+    fn default() -> Self {
+        Self::new(vec![])
+    }
+}
+
+/// Timespan with a fixed duration, recurring on weekly basis.
+///
+/// Length duration is capped so that a window never crosses week boundary.
+#[derive(Clone, Debug, Eq, Serialize)]
+pub struct WeeklyWindow {
+    start_day: chrono::Weekday,
+    start_hour: u8,
+    start_minute: u8,
+    length: Duration,
+}
+
+impl WeeklyWindow {
+    /// Parse a timespan into weekly windows.
+    pub fn parse_timespan(
+        start_day: chrono::Weekday,
+        start_hour: u8,
+        start_minute: u8,
+        length: Duration,
+    ) -> Fallible<Vec<Self>> {
+        failure::ensure!(
+            start_hour <= 24 && start_minute <= 59,
+            "invalid start time: {}:{}",
+            start_hour,
+            start_minute
+        );
+
+        let mut windows = vec![];
+        let mut remaining_length = utils::check_duration(length)?;
+        let mut start_weekly_minute =
+            utils::time_as_weekly_minute(start_day, start_hour, start_minute);
+        while remaining_length.as_secs() > 0 {
```

Comment:
This is an interesting construction. Really this can iterate at most twice, right? It might be clearer to just change `chop_timestan` to return a Vec of either one or two windows.

Reply:
Good catch! Both the while-loop and most of `chop_timespan` logic are not actually useful anymore after moving away from daily windows. I moved some code around and dropped both.